### PR TITLE
Polling and Playback Framework

### LINKED ISSUE #21: Queueing/Playing Framework
Class outline for how we handle all this stuff

### PR DESCRIPTION
**Pull request by daniel-centore**
Creates a structure within which we can communicate with the server and update the UI and playback status

Resolves #21

### CODE REVIEW

File: MeNext/Playing/MainController.cs
Comment (ConnorFoody) [inline MeNext/Playing/MainController.cs:121]: Indentation

File: MeNext/Playing/Poller.cs
Comment (ConnorFoody) [inline MeNext/Playing/Poller.cs:10]: Do we want to make this variable?

File: MeNext/Playing/Poller.cs
Comment (ConnorFoody) [inline MeNext/Playing/Poller.cs:47]: Why on main thread?

File: MeNext/Playing/Poller.cs
Comment (ConnorFoody) [inline MeNext/Playing/Poller.cs:31]: Will we pull this out once testing is over? Where do we track the changeID?

File: MeNext/Playing/MainController.cs
Comment (ConnorFoody) [inline MeNext/Playing/MainController.cs:127]: brackets

File: Droid/Playing/PollingService.cs
Comment (ConnorFoody) [inline Droid/Playing/PollingService.cs:15]: Why can't we have the shared code spin off it's own thread / service?

File: MeNext/Playing/MainController.cs
Comment (ConnorFoody) [inline MeNext/Playing/MainController.cs:370]: Do we want to provide some mechanism to pre-fetch the next song?

File: MeNext/Playing/PlayController.cs
Comment (ConnorFoody) [inline MeNext/Playing/PlayController.cs:39]: Don't we need to factor in trip time?

File: MeNext/Playing/MainController.cs
Comment (ConnorFoody) [inline MeNext/Playing/MainController.cs:108]: brackets, indentation

File: MeNext/Playing/MainController.cs
Comment (ConnorFoody) [inline MeNext/Playing/MainController.cs:335]: indentation

File: MeNext/Playing/MainController.cs
Comment (ConnorFoody) [inline MeNext/Playing/MainController.cs:354]: Where does the state get passed to the other controllers?

File: MeNext/Playing/MainController.cs
Comment (ConnorFoody) [inline MeNext/Playing/MainController.cs:366]: Why do we need to know which song ended?

File: MeNext/Playing/MainController.cs
Comment (ConnorFoody) [inline MeNext/Playing/MainController.cs:350]: People other than the host need this info, right?

**Review by ConnorFoody — COMMENTED** [review]
On the formatting front: 
1. Some indentation issues, I pointed a few out
2. I like brackets around if statements



Its not clear why we need to have platform specific implementations for starting background tasks. It seems like we should be able to do it in pure xamarin forms. 

How will info be propogated through to models / controllers? Do they all subscribe throught the messaging center? Is the message parsed and pushed through the music controller?

**Comment by ConnorFoody** [comment]
@mluquette I think that this may be relevant for you...

File: MeNext/Playing/MainController.cs
Comment (daniel-centore) [reply MeNext/Playing/MainController.cs:350]: Yes (see the incredibly specific TODO on line 345). `playController` controls _actually_ playing the music, which only the host does.

File: MeNext/Playing/MainController.cs
Comment (daniel-centore) [reply MeNext/Playing/MainController.cs:366]: We don't. It's part of the interface for other potential users of the music service.

File: MeNext/Playing/MainController.cs
Comment (daniel-centore) [reply MeNext/Playing/MainController.cs:370]: It doesn't work in Spotify yet so very low priority

File: MeNext/Playing/PlayController.cs
Comment (daniel-centore) [reply MeNext/Playing/PlayController.cs:39]: The trip time correction needs to be done earlier in the process, _ie_ when the communication itself happens, so we can time how long it takes. The status message is passed thru the message handler so it doesn't know how long the trip took.

File: MeNext/Playing/MainController.cs
Comment (daniel-centore) [reply MeNext/Playing/MainController.cs:354]: In SomethingChanged(), which is not yet written

File: MeNext/Playing/Poller.cs
Comment (daniel-centore) [reply MeNext/Playing/Poller.cs:10]: Yes

File: MeNext/Playing/Poller.cs
Comment (daniel-centore) [reply MeNext/Playing/Poller.cs:31]: This will be pulled out, yes. `ChangeId` will be extracted by when we "obtain a real status message" on line 39.

File: MeNext/Playing/Poller.cs
Comment (daniel-centore) [reply MeNext/Playing/Poller.cs:47]: Because that's what the tutorial did :)

**Comment by daniel-centore** [comment]
> Some indentation issues, I pointed a few out

It would appear that Xamarin ended up mixing spaces and tabs for indentation. Ugh. Will fix

> I like brackets around if statements

I either use K&R or allow 1 line if statements. I'll switch the project settings to K&R.

> Its not clear why we need to have platform specific implementations for starting background tasks. It seems like we should be able to do it in pure xamarin forms.

"Unfortunately, the way that these two platforms have evolved to handle executing background code is completely different. As such, there is no way that we can abstract the backgrounding feature into the Xamarin.Forms library." - [Source](http://arteksoftware.com/backgrounding-with-xamarin-forms/)

> How will info be propogated through to models / controllers? Do they all subscribe throught the messaging center? Is the message parsed and pushed through the music controller?

UI updates will be done from `SomethingChanged()`. Playback info is already propagated to `UpdateActualPlaying` in `PlayController`.

File: MeNext/Playing/PlayController.cs
Comment (ConnorFoody) [reply MeNext/Playing/PlayController.cs:39]: OK, I guess we'll put that in the poller later?

File: MeNext/Playing/MainController.cs
Comment (ConnorFoody) [reply MeNext/Playing/MainController.cs:354]: OK, doesn't look like SomethingChanged has a param? Maybe pass message to it now?

**Comment by daniel-centore** [comment]
Actually, putting off K&R switch until all branches are merged into dev.

**Review by ConnorFoody — APPROVED** [review]
LGTM, I guess we'll catch more specifics down the road? 

Also, not that I'm good at this, but lets try to break stuff down into multiple PRs.

File: Droid/Playing/PollingService.cs
Comment (daniel-centore) [reply Droid/Playing/PollingService.cs:15]: Don't understand the question

File: MeNext/Playing/MainController.cs
Comment (daniel-centore) [reply MeNext/Playing/MainController.cs:354]: I figured we'd add the req'd params when we impl it

File: MeNext/Playing/PlayController.cs
Comment (daniel-centore) [reply MeNext/Playing/PlayController.cs:39]: Indeed. Blocked on the @mluquette project to do that.

**Comment by daniel-centore** [comment]
> LGTM, I guess we'll catch more specifics down the road?

Yep

> Also, not that I'm good at this, but lets try to break stuff down into multiple PRs.

Sorry, this was a hard one to split up.

File: Droid/Playing/PollingService.cs
Comment (ConnorFoody) [reply Droid/Playing/PollingService.cs:15]: It seems like we can do something like [forms tasks](https://developer.xamarin.com/api/type/System.Threading.Tasks.Task/) in the shared code to start background jobs there...

Comment (daniel-centore) [reply Droid/Playing/PollingService.cs:15]: Forms tasks only work while the app is open, so the host wouldn't be able to continue playing music if the app were closed. Only the native tasks have this flexibility (even though I haven't actually configured it yet).

Comment (ConnorFoody) [reply Droid/Playing/PollingService.cs:15]: bummer, at least its not too much extra work to get a background task up

**Pull request merged**
